generator/internal/parser/httprule/http_rule_parser.go

```diff
@@ -0,0 +1,382 @@
+type PathTemplate struct {
+	Segments []*Segment
+	Verb     *Literal
+}
+
+// Match represents a single '*' match.
+type Match struct{}
+
+// MatchRecursive represents a '**' match.
+type MatchRecursive struct{}
+
+type Literal string
+type Identifier string
+
+// Variable represents a variable in the path template with its field path and nested segments.
+type Variable struct {
+	FieldPath []*Identifier
+	Segments  []*Segment
+}
+
+// Segment represents a single segment of the path template, which can hold one of several types of values.
+type Segment struct {
+	Literal        *Literal
+	Match          *Match
+	MatchRecursive *MatchRecursive
+	Variable       *Variable
+}
+
+const eof = -1
+
+const (
+	slash    = '/'
+	star     = '*'
+	varLeft  = '{'
+	varRight = '}'
+	varSep   = '='
+	identSep = '.'
+	verbSep  = ':'
+)
+
+func parsePathTemplate(pathTemplate string) (*PathTemplate, error) {
+	if len(pathTemplate) < 2 {
+		return nil, fmt.Errorf("invalid path template, expected at least two characters: %s", pathTemplate)
+	} else if pathTemplate[0] != slash {
+		return nil, fmt.Errorf("invalid path template, expected it to start with '/': %s", pathTemplate)
+	} else if pathTemplate[len(pathTemplate)-1] == slash {
+		return nil, fmt.Errorf("invalid path template, expected it to not end with '/': %s", pathTemplate)
+	}
+
+	lastPos := len(pathTemplate)
+	var err error
+	var verb *Literal
+	if strings.ContainsRune(pathTemplate, verbSep) {
+		lastPos = strings.LastIndex(pathTemplate, string(verbSep))
+		verb, err = parseVerb(pathTemplate[lastPos:])
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	segments, err := parseSegments(pathTemplate[1:lastPos])
+	if err != nil {
+		return nil, err
+	}
+
+	return &PathTemplate{
+		Segments: segments,
+		Verb:     verb,
+	}, nil
+
+}
+
+func parseVerb(verbString string) (*Literal, error) {
+	if len(verbString) == 0 {
+		return nil, nil
+	}
+	if len(verbString) < 2 {
+		return nil, fmt.Errorf("invalid verb, when not empty, must have at least two characters")
+	} else if verbString[0] != verbSep {
+		return nil, fmt.Errorf("invalid verb, must start with '%q': %s", verbSep, verbString)
+	}
+	return parseLiteral(verbString[1:])
+}
+
+// parseSegments parses the first segment out of the provided string, and calls itself recursively to parse the remaining segments, if any.
+func parseSegments(segmentsString string) ([]*Segment, error) {
+	if len(segmentsString) < 1 {
+		return nil, fmt.Errorf("invalid segments, expected at least one character")
+	} else if segmentsString[0] == slash {
+		return nil, fmt.Errorf("invalid segments, cannot start with '/': %s", segmentsString)
+	} else if segmentsString[len(segmentsString)-1] == slash {
+		return nil, fmt.Errorf("invalid segments, cannot end with '/': %s", segmentsString)
+	}
+
+	var firstSegment *Segment
+	var lastPos int
+	var err error
+	if segmentsString[0] == varLeft {
+		lastPos = strings.Index(segmentsString, string(varRight))
```

Comment:
I think you want `strings.LastIndex` here?  Alternatively, you can change these `parse*()` functions to return where they stopped parsing, and (in this case) require that `parseVarSegment` stops on a `}`.

Reply:
I was trying to model the functions the way the syntax describes them, so the `parseSegments` expects at least one segment to exist, parses it, and if more segments are present, calls itself again. But the other functions act only on the segments that are given to them and expect their segment to be perfectly aligned with the end of the string. 
It isn't the most efficient parsing algorithm, as `parseSegments` does a lot of array contacting and string slicing, but I was mostly interested in keeping the code structure very close to the syntax.

Reply:
If all your functions return "I got these elements and this is what remains to be parsed" then the caller can check:

1) The callee stopped at a `/` I need to recurse.
2) The callee stopped at a `:` maybe there is a trailing Verb.
3) I expected the callee to parse the whole thing, but there are some left overs (bad chars or something) that is an error.
4) I expected the callee to consume everything until the matching `}`, that is an error.

For example, the top-level function becomes

```golang
func parsePathTemplate(input string) (*pathTemplate, error) {
  pathTemplate := &PathTemplate{Segments: segments}  
  unparsed, err := parseSegments(pathTemplate, input)
  if err != nil {
    return nil, fmt.Errorf("%w, input=%q, err, input)
  }
  unparsed, err = parseOptionalVerb(pathTemplate, unparsed)
  if err != nil {
    return nil, fmt.Errorf("%w, input=%q, err, input)
  }
  return pathTemplate
}

func parseOptionalVerb(pathTemplate *PathTemplate, input string) (string, error) {
  if input == "" {
    return "", nil
  }
  if !strings.StartsWith(input, ":") {
    return input, fmt.Errorf("error parsing http rule, expected ':' to start VERB")
  }
  pathTemplate.Verb = input[1:]
  return "", nil
}

func parseSegments(pathTemplate *PathTemplate, input string) (string, error) {
  unparsed, err := parseSegment(pathTemplate, input)
  if err != nil {
      return "", err
  }
  while strings.StartsWith(unparsed, "/") {
    unparsed, err = parseSegment(pathTemplate, unparsed[1:])
    if err != nil {
      return "", err
    }
  }
  if len(pathTemplate.Segments) == 0 {
    return "", fmt.Errorf("the path segments should not be empty")    
  }
  return unparsed, nil
}
```

Reply:
PS: If I am getting nit picky, feel free to tell me so, we can improve the code in multiple iterations.